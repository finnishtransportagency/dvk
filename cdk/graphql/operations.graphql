query findFairways {
  fairways {
    ...fragment FairwayFragment on Fairway {
      id
      nameFI
      nameSV
    }
  }
}

query findFairway($id: Int) {
  fairway(id: $id) {
    ...Fairway
  }
}
