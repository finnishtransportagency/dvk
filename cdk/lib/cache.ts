export type CloudFrontCacheKey =
  | 'aislocations'
  | 'aisvessels'
  | 'mareograph'
  | 'observation'
  | 'buoy'
  | 'marinewarning'
  | 'feature'
  | 'pilotroutes'
  | 'forecast';

export const FEATURE_CACHE_DURATION = 7200; // 2 hours

type CloudFrontCacheItem = {
  key: CloudFrontCacheKey;
  maxAge: number;
  staleWhileRevalidate: number;
  staleIfError: number;
};

const CloudFrontCacheItems: Array<CloudFrontCacheItem> = [
  {
    key: 'aislocations',
    maxAge: 2, // 2 seconds
    staleWhileRevalidate: 3, // 3 seconds
    staleIfError: 58, // 58 seconds (do not return location data more than 1 minute old)
  },
  {
    key: 'aisvessels',
    maxAge: 300, // 5 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 300, // 5 minutes (do not return vessel data more than 10 minutes old)
  },
  {
    key: 'mareograph',
    maxAge: 60, // 1 minute
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'observation',
    maxAge: 60, // 1 minute
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'buoy',
    maxAge: 60, // 1 minute
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'marinewarning',
    maxAge: 540, // 9 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'feature',
    maxAge: 7140, // 1 hour 59 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'pilotroutes',
    maxAge: 7140, // 1 hour 59 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
  {
    key: 'forecast',
    maxAge: 7140, // 1 hour 59 minutes
    staleWhileRevalidate: 60, // 1 minute
    staleIfError: 12 * 3600, // 12 hours
  },
];

export function getAisCacheControlHeaders(key: 'aisvessels' | 'aislocations'): Record<string, string[]> {
  return getCloudFrontCacheControlHeaders(key);
}

export function isCloudFrontCacheKey(key: string): key is CloudFrontCacheKey {
  return CloudFrontCacheItems.some((item) => item.key === key );
}

export function getFeatureCacheControlHeaders(): Record<string, string[]> {
  return getCloudFrontCacheControlHeaders('feature');
}

export function getCloudFrontCacheControlHeaders(key: CloudFrontCacheKey): Record<string, string[]> {
  const cacheItem = CloudFrontCacheItems.find((item) => item.key === key );
  if (!cacheItem) {
    return getFeatureCacheControlHeaders();
  }
  const maxAge = cacheItem.maxAge;
  const staleWhileRevalidate = cacheItem.staleWhileRevalidate;
  const staleIfError = cacheItem.staleIfError;

  return {
    'Cache-Control': ['max-age=' + maxAge + ', ' + 'stale-while-revalidate=' + staleWhileRevalidate + ', ' + 'stale-if-error=' + staleIfError],
  };
}
