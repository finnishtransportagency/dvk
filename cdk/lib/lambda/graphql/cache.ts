import { DeleteObjectsCommand, GetObjectCommand, ObjectIdentifier, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { getEnvironment } from '../environment';
import { log } from '../logger';
import { Readable } from 'stream';

const s3Client = new S3Client({ region: 'eu-west-1' });

const AIS_LOCATION_CACHE = {
  MAX_AGE: 2, // 2 seconds
  STALE_WHILE_REVALIDATE: 3, // 3 seconds
  STALE_IF_ERROR: 58, // 58 seconds (do not return location data more than 1 minute old)
};
const AIS_VESSEL_CACHE = {
  MAX_AGE: 300, // 5 minutes
  STALE_WHILE_REVALIDATE: 60, // 1 minute
  STALE_IF_ERROR: 300, // 5 minutes (do not return vessel data more than 10 minutes old)
};

export const FEATURE_CACHE_DURATION = 7200; // 2 hours

function getCacheBucketName() {
  return `featurecache-${getEnvironment()}`;
}

export function getFeatureCacheDuration(key: string) {
  if (key === 'aislocations') {
    return AIS_LOCATION_CACHE.MAX_AGE;
  } else if (key === 'aisvessels') {
    return AIS_VESSEL_CACHE.MAX_AGE;
  } else {
    return FEATURE_CACHE_DURATION;
  }
}

export function getCacheControlHeaders(key: string): Record<string, string[]> {
  if (key === 'aislocations') {
    return {
      'Cache-Control': [
        'max-age=' +
          AIS_LOCATION_CACHE.MAX_AGE +
          ', ' +
          'stale-while-revalidate=' +
          AIS_LOCATION_CACHE.STALE_WHILE_REVALIDATE +
          ', ' +
          'stale-if-error=' +
          AIS_LOCATION_CACHE.STALE_IF_ERROR,
      ],
    };
  } else if (key === 'aisvessels') {
    return {
      'Cache-Control': [
        'max-age=' +
          AIS_VESSEL_CACHE.MAX_AGE +
          ', ' +
          'stale-while-revalidate=' +
          AIS_VESSEL_CACHE.STALE_WHILE_REVALIDATE +
          ', ' +
          'stale-if-error=' +
          AIS_VESSEL_CACHE.STALE_IF_ERROR,
      ],
    };
  } else {
    return {};
  }
}

export async function cacheResponse(key: string, response: object | string) {
  const cacheDurationSeconds = getFeatureCacheDuration(key);
  const expires = new Date();
  expires.setTime(expires.getTime() + cacheDurationSeconds * 1000);
  const command = new PutObjectCommand({
    Key: key,
    Bucket: getCacheBucketName(),
    Expires: expires,
    Body: typeof response === 'string' ? response : JSON.stringify(response),
  });
  await s3Client.send(command);
}

export type CacheResponse = {
  expired: boolean;
  data?: string;
  lastModified?: string;
};

export async function getFromCache(key: string): Promise<CacheResponse> {
  try {
    const data = await s3Client.send(
      new GetObjectCommand({
        Key: key,
        Bucket: getCacheBucketName(),
      })
    );
    if (data.Body) {
      log.debug(`returning ${key} from cache`);
      const streamToString = (stream: Readable): Promise<string> =>
        new Promise((resolve, reject) => {
          const chunks: Uint8Array[] = [];
          stream.on('data', (chunk: Uint8Array) => chunks.push(chunk));
          stream.on('error', reject);
          stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
        });
      const expires = data.ExpiresString ? Date.parse(data.ExpiresString) : undefined;
      return {
        expired: expires !== undefined && !isNaN(expires) && expires < Date.now(),
        data: await streamToString(data.Body as Readable),
        lastModified: data.LastModified?.toUTCString(),
      };
    }
  } catch (e) {
    // errors ignored also not found
  }
  return { expired: true };
}

export async function deleteCacheObjects(keys: string[]) {
  const objects: ObjectIdentifier[] = keys.map((key) => {
    return { Key: key };
  });
  const response = await s3Client.send(
    new DeleteObjectsCommand({
      Bucket: getCacheBucketName(),
      Delete: {
        Objects: objects,
      },
    })
  );
  log.debug(`Deleted cache objects ${keys}: ${response.Deleted?.length}/${keys.length}`);
  if (response.Errors?.length) {
    log.error(response.Errors);
  }
}
