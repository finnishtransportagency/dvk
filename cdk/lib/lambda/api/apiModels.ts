import { Geometry } from 'geojson';
import { GeometryPoint, Text } from '../../../graphql/generated';
import { FairwayCardIdName } from '../db/fairwayCardDBModel';

export type VaylaGeojsonFeature = { type?: string; geometry: Geometry };

/* VATU */
type MitoitusAlusAPIModel = {
  alustyyppiKoodi: string;
  alustyyppi: string;
  pituus: number;
  leveys: number;
  syvays: number;
  koko?: number;
  runkoTaytelaisyysKerroin?: number;
};

type LuokitusAPIModel = {
  luokitusTyyppi: string;
  vaylaluokkaKoodi: string;
  vaylaluokkaFI?: string;
  vaylaluokkaSV?: string;
};

export type VaylaAPIModel = {
  jnro: number;
  nimiFI: string;
  nimiSV?: string;
  vaylalajiKoodi?: string;
  vaylaLajiFI?: string;
  vaylaLajiSV?: string;
  valaistusKoodi?: string;
  valaistusFI?: string;
  valaistusSV?: string;
  omistaja?: string;
  merialueFI?: string;
  merialueSV?: string;
  alunSeloste?: string;
  paatepisteenSeloste?: string;
  normaaliKaantosade?: number;
  minimiKaantosade?: number;
  normaaliLeveys?: number;
  minimiLeveys?: number;
  varavesi?: string;
  lisatieto?: string;
  mareografi?: string;
  mitoitusalus?: MitoitusAlusAPIModel[];
  luokitus?: LuokitusAPIModel[];
};

export type VaylaFeature = { properties: VaylaAPIModel };
export type VaylaFeatureCollection = { type?: string; features: VaylaFeature[] };

type AlueVaylaAPIModel = {
  jnro: number;
  status: number;
  linjaus: number;
  mitoitusNopeus?: number;
  mitoitusNopeus2?: number;
  vaylaalueenJarjestysNro?: number;
  nimiFI?: string;
  nimiSV?: string;
};

type AlueProperties = {
  id: number;
  nimi?: string;
  mitoitusSyvays?: number;
  harausSyvyys?: number;
  vertaustaso?: string;
  n2000MitoitusSyvays?: number;
  n2000HarausSyvyys?: number;
  n2000Vertaustaso?: string;
  suunta?: number;
  diaariNumero?: string;
  vahvistusPaivamaara?: string;
  omistaja?: string;
  lisatieto?: string;
  vayla?: AlueVaylaAPIModel[];
  tyyppiKoodi?: number;
  tyyppi?: string;
  merkintalajiKoodi?: number;
  merkintalaji?: string;
  liikennointiStatusKoodi?: number;
  liikennointiStatus?: string;
  liikennointiTyyppiKoodi?: number;
  liikenteenTyyppi?: string;
  liikennointiSuuntaKoodi?: number;
  liikennointiSuunta?: string;
};

export type AlueFeature = {
  properties: AlueProperties;
} & VaylaGeojsonFeature;

export type AlueFeatureCollection = {
  type?: string;
  features: AlueFeature[];
};

type NavigointiLinjaVaylaAPIModel = {
  jnro: number;
  status?: number;
  linjaus?: number;
  nimiFI?: string;
  nimiSV?: string;
};

type NavigointiLinjaProperties = {
  id: number;
  mitoitusSyvays?: number;
  harausSyvyys?: number;
  vertaustaso?: string;
  n2000MitoitusSyvays?: number;
  n2000HarausSyvyys?: number;
  n2000Vertaustaso?: string;
  tosisuunta?: number;
  pituus?: number;
  diaariNumero?: string;
  vahvistusPaivamaara?: string;
  omistaja?: string;
  lisatieto?: string;
  tyyppiKoodi?: string;
  tyyppi?: string;
  vayla: NavigointiLinjaVaylaAPIModel[];
};

export type NavigointiLinjaFeature = {
  properties: NavigointiLinjaProperties;
} & VaylaGeojsonFeature;

export type NavigointiLinjaFeatureCollection = {
  type?: string;
  features: NavigointiLinjaFeature[];
};

type RajoitustyyppiAPIModel = {
  koodi?: string;
  rajoitustyyppi?: string;
};

type RajoitusVaylaAPIModel = {
  jnro: number;
  nimiFI?: string;
  nimiSV?: string;
};

type RajoitusAlueProperties = {
  id: number;
  rajoitustyyppi?: string;
  rajoitustyypit?: RajoitustyyppiAPIModel[];
  suuruus?: number;
  esittaja?: string;
  diaariNumero?: string;
  vahvistusPaivamaara?: string;
  muutosPaivamaara?: string;
  alkuPaivamaara?: string;
  loppuPaivamaara?: string;
  paatosTila?: string;
  tietolahde?: string;
  sijainti?: string;
  kunta?: string;
  poikkeus?: string;
  vayla?: RajoitusVaylaAPIModel[];
};

export type RajoitusAlueFeature = { properties: RajoitusAlueProperties } & VaylaGeojsonFeature;

export type RajoitusAlueFeatureCollection = {
  type?: string;
  features: RajoitusAlueFeature[];
};

export type TurvalaiteVaylaAPIModel = {
  jnro: number;
  paavayla?: string;
};

export type TurvalaiteVikatiedotProperties = {
  vikaId: number;
  turvalaiteNumero: number;
  turvalaiteNimiFI: string;
  turvalaiteNimiSV?: string;
  vikatyyppiKoodi: number;
  vikatyyppiFI: string;
  vikatyyppiSV?: string;
  vikatyyppiEN?: string;
  kirjausAika: string;
};
export type TurvalaiteVikatiedotFeature = { properties: TurvalaiteVikatiedotProperties } & VaylaGeojsonFeature;
export type TurvalaiteVikatiedotFeatureCollection = {
  type?: string;
  features: TurvalaiteVikatiedotFeature[];
};

type TurvalaiteReunaetaisyysAPIModel = {
  vaylaalueID: number;
  etaisyys: number;
};

type TurvalaiteProperties = {
  turvalaitenumero: number;
  nimiFI?: string;
  nimiSV?: string;
  alityyppi?: string;
  turvalaitetyyppiKoodi?: number;
  turvalaitetyyppiFI?: string;
  turvalaitetyyppiSV?: string;
  navigointilajiKoodi?: number;
  navigointilajiFI?: string;
  navigointilajiSV?: string;
  valaistu?: string;
  omistajaFI?: string;
  omistajaSV?: string;
  symboli?: string;
  kaukohallinta?: number;
  AISTyyppi?: number;
  AISTyyppiSeliteFI?: string;
  AISTyyppiSeliteSV?: string;
  vayla?: TurvalaiteVaylaAPIModel[];
  reunaetaisyys?: TurvalaiteReunaetaisyysAPIModel[];
};

export type TurvalaiteFeature = { properties: TurvalaiteProperties } & VaylaGeojsonFeature;

export type TurvalaiteFeatureCollection = {
  type?: string;
  features: TurvalaiteFeature[];
};

export type TaululinjaVaylaAPIModel = {
  jnro: number;
  nimiFI?: string;
  nimiSV?: string;
};

export type TaululinjaProperties = {
  taululinjaId: number;
  suunta: number;
  etuTLId?: number;
  keskiTLId?: number;
  takaTLId?: number;
  vayla: TaululinjaVaylaAPIModel[];
};

export type TaululinjaFeature = { properties: TaululinjaProperties } & VaylaGeojsonFeature;
export type TaululinjaFeatureCollection = {
  type?: string;
  features: TaululinjaFeature[];
};

type KaantoympyraVaylaAPIModel = {
  jnro: number;
  nimiFI: string;
  nimiSV?: string;
  luokitus: number;
};

export type KaantoympyraProperties = {
  kaantoympyraID: number;
  halkaisija: number;
  haraussyvyys?: number;
  vertaustaso?: string;
  lisatieto?: string;
  vayla?: KaantoympyraVaylaAPIModel[];
};

export type KaantoympyraFeature = { properties: KaantoympyraProperties } & VaylaGeojsonFeature;

export type KaantoympyraFeatureCollection = { type?: string; features: KaantoympyraFeature[] };

/* Traficom */
export type PilotPlace = {
  id: number;
  name: Text;
  geometry: GeometryPoint;
  fairwayCards: FairwayCardIdName[];
};

/* Pooki */
export type MarineWarningDates = {
  startDateTime?: number;
  endDateTime?: number;
  dateTime?: number;
};

/* Weather */
export type WeatherMareograph = {
  fmisid: number;
  geoid: number;
  latlon: string;
  station_name: string;
  localtime: string;
  WLEV_PT1S_INSTANT: number;
  WLEVN2K_PT1S_INSTANT: number;
};

export type Mareograph = {
  id: number;
  geometry: Geometry;
  name: string;
  dateTime: number;
  waterLevel: number;
  n2000WaterLevel: number;
  calculated: boolean;
};

export type WeatherObservation = {
  fmisid: number;
  geoid: number;
  latlon: string;
  stationname: string;
  localtime: string;
  WG_PT10M_MAX: number;
  WD_PT10M_AVG: number;
  WS_PT10M_AVG: number;
  TA_PT1M_AVG: number;
  VIS_PT1M_AVG: number | null;
};

export type Observation = {
  id: number;
  name: string;
  windSpeedAvg: number;
  windSpeedMax: number;
  windDirection: number;
  visibility: number | null;
  temperature: number;
  dateTime: number;
  geometry: Geometry;
};

export type WeatherBuoy = {
  fmisid: number;
  geoid: number;
  latlon: string;
  station_name: string;
  localtime: string;
  WH_PT1M_ACC: number | null;
  WHD_PT1M_ACC: number | null;
  TW_PT1M_AVG: number | null;
};

export type Buoy = {
  id: number;
  name: string;
  dateTime: number;
  geometry: Geometry;
  temperature: number | null;
  waveDirection: number | null;
  waveHeight: number | null;
};

/* AIS */
export type VesselAPIModel = {
  name: string;
  timestamp: number;
  mmsi: number;
  callSign: string;
  imo: number;
  shipType: number;
  draught: number;
  eta: number;
  posType: number;
  referencePointA: number;
  referencePointB: number;
  referencePointC: number;
  referencePointD: number;
  destination: string;
};

type VesselLocationProperties = {
  mmsi: number;
  sog: number;
  cog: number;
  navStat: number;
  rot: number;
  posAcc: boolean;
  raim: boolean;
  heading?: number;
  timestamp: number;
  timestampExternal: number;
};

type VesselLocationFeature = {
  mmsi: number;
  type?: string;
  geometry: Geometry;
  properties: VesselLocationProperties;
};

export type VesselLocationFeatureCollection = {
  type?: string;
  dataUpdatedTime?: string;
  features: VesselLocationFeature[];
};

export type Vessel = {
  name: string;
  timestamp: Date;
  mmsi: number;
  callSign: string;
  imo: number;
  shipType: number;
  draught: number; // meters
  eta: string; // MM-DD HH:mm
  posType: number;
  referencePointA: number;
  referencePointB: number;
  referencePointC: number;
  referencePointD: number;
  destination: string;
};

/* RTZ */
export type Coordinate = [number, number];

export type RtzWaypoint = {
  coordinate: Coordinate;
  turnRadius: number;
};

type RtzGeometria = {
  type: string;
  coordinates: Coordinate;
};

export type RtzReittipiste = {
  tunnus: number;
  nimi: string;
  rtzTunniste: number;
  reittitunnus: number;
  kaarresade: number;
  geometria: RtzGeometria;
  leveysVasen: number;
  leveysOikea: number;
  geometriaTyyppi: string;
  muutosaikaleima: string;
  jarjestys: number;
};

export type RtzData = {
  tunnus: number;
  tila: number;
  nimi: string;
  tunniste: string;
  rtz: string;
  reittipisteet: Array<RtzReittipiste>;
};

/* IBNet */
export type IBNetApiResponse = {
  rvEndpoints: string[];
  toRv: string;
};

export type Dirway = {
  id: string;
  name: string;
  description: string;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};

export type DirwayPoint = {
  id: string;
  dirway_id: string;
  order_num: number;
  name: string;
  latitude: number;
  longitude: number;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};

export type LocationApiModel = {
  id: string;
  name: string;
  type: 'PORT' | 'FAIRWAY';
  locode_list: string;
  nationality: string;
  latitude: number;
  longitude: number;
  winterport: boolean;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};

export type RestrictionApiModel = {
  id: string;
  location_id: string;
  start_time: string;
  end_time?: string | null;
  text_compilation: string;
  rv: number | null;
  change_time: string;
  deleted?: boolean | null;
};
