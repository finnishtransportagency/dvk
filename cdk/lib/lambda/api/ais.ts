import { Feature, FeatureCollection, GeoJsonProperties, Geometry } from 'geojson';
import { Vessel, VesselAPIModel, VesselLocationFeatureCollection } from './apiModels';
import { fetchAISFeatureCollection, fetchAISMetadata } from './axios';

const aisV1Path = 'ais/api/ais/v1/';

function getTimestampParams(): Record<string, string> {
  const now = Date.now();
  const yesterday = Date.now() - 24 * 60 * 60 * 1000;
  return { from: yesterday.toString(), to: now.toString() };
}

function parseFrom64BitInteger(integer: number, bitStart: number, bitEnd: number) {
  // 64-bit integer: get bit range from start to end by losing other bits
  // Shift bits first to left and then to right, let others fall off
  const i = (integer << (63 - bitStart)) >>> (63 - bitStart + bitEnd);
  return i.toString().padStart(2, '0');
}

/* MMDDHHMM UTC
Bits 19-16: month; 1-12; 0 = not available = default
Bits 15-11: day; 1-31; 0 = not available = default
Bits 10-6: hour; 0-23; 24 = not available = default
Bits 5-0: minute; 0-59; 60 = not available = default
*/
function parseETA(eta: number): string {
  const month = parseFrom64BitInteger(eta, 19, 16);
  const day = parseFrom64BitInteger(eta, 15, 11);
  const hour = parseFrom64BitInteger(eta, 10, 6);
  const minute = parseFrom64BitInteger(eta, 5, 0);
  return month + '-' + day + ' ' + hour + ':' + minute;
}

export async function fetchVessels(): Promise<Vessel[]> {
  const path = aisV1Path + 'vessels';
  const params = getTimestampParams();
  const data: VesselAPIModel[] = await fetchAISMetadata(path, params);
  return data.map((row) => {
    return {
      name: row.name,
      timestamp: new Date(row.timestamp),
      mmsi: row.mmsi,
      callSign: row.callSign,
      imo: row.imo,
      shipType: row.shipType,
      draught: row.draught / 10,
      eta: parseETA(row.eta),
      posType: row.posType,
      referencePointA: row.referencePointA,
      referencePointB: row.referencePointB,
      referencePointC: row.referencePointC,
      referencePointD: row.referencePointD,
      destination: row.destination,
    };
  });
}

export async function fetchLocations(): Promise<FeatureCollection> {
  const path = aisV1Path + 'locations';
  const params = getTimestampParams();
  const data: VesselLocationFeatureCollection = await fetchAISFeatureCollection(path, params);
  const { dataUpdatedTime, features } = data;
  const geojsonFeatures: Feature<Geometry, GeoJsonProperties>[] = features.map((f) => {
    return {
      type: 'Feature',
      id: f.mmsi,
      geometry: f.geometry,
      properties: {
        ...f.properties,
        dataUpdatedTime,
        timestampExternal: new Date(f.properties.timestampExternal),
        featureType: 'aisvessel',
      },
    };
  });
  const collection: FeatureCollection = {
    type: 'FeatureCollection',
    features: geojsonFeatures,
  };
  return collection;
}
