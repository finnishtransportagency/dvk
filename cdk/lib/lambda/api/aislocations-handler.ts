import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { fetchLocations } from './ais';
import { handleAisCall } from '../util';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `aislocations()`);
  return handleAisCall('aislocations', fetchLocations, ['application/geo+json']);
};
