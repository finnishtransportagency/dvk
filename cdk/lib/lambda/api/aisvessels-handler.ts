import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { fetchVessels } from './ais';
import { handleAisCall } from '../util';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `aisvessels()`);
  return handleAisCall('aisvessels', fetchVessels, ['application/json']);
};
