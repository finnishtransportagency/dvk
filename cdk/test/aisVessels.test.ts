import { handler } from '../lib/lambda/api/aisvessels-handler';
import { mockAISALBEvent } from './mocks';
import { gunzip } from 'zlib';
import assert from 'assert';
import { Vessel, VesselAPIModel } from '../lib/lambda/api/apiModels';
import { getCacheControlHeaders } from '../lib/lambda/graphql/cache';

const path = 'aisvessels';

jest.mock('../lib/lambda/environment', () => ({
  getEnvironment: () => 'mock',
  isPermanentEnvironment: () => false,
  getHeaders: () => {},
}));

const vessels: VesselAPIModel[] = [
  {
    name: 'MOCK1',
    timestamp: 1650652205774,
    mmsi: 11122233001,
    callSign: 'MC001',
    imo: 7734789,
    shipType: 52,
    draught: 32,
    eta: 210112,
    posType: 1,
    referencePointA: 8,
    referencePointB: 20,
    referencePointC: 4,
    referencePointD: 4,
    destination: 'KOTKA',
  },
  {
    name: 'MOCK2',
    timestamp: 1644101972418,
    mmsi: 99988877001,
    callSign: 'MC002',
    imo: 9505508,
    shipType: 90,
    draught: 75,
    eta: 219136,
    posType: 1,
    referencePointA: 35,
    referencePointB: 63,
    referencePointC: 10,
    referencePointD: 10,
    destination: 'TURKU',
  },
  {
    name: 'MOCK3',
    timestamp: 1659503493045,
    mmsi: 66655544001,
    callSign: 'MC003',
    imo: 9684823,
    shipType: 52,
    draught: 56,
    eta: 129536,
    posType: 3,
    referencePointA: 12,
    referencePointB: 12,
    referencePointC: 6,
    referencePointD: 5,
    destination: 'HELSINKI',
  },
];

async function parseResponse(body: string): Promise<Vessel[]> {
  const response = new Promise<Error | Buffer>((resolve, reject) =>
    gunzip(Buffer.from(body, 'base64'), (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(data);
    })
  );
  return JSON.parse((await response).toString()) as Vessel[];
}

let throwError = false;
jest.mock('../lib/lambda/api/axios', () => ({
  fetchAISMetadata: () => {
    if (throwError) {
      throw new Error('Fetching from AIS api failed');
    }
    return vessels;
  },
}));

beforeEach(() => {
  jest.resetAllMocks();
  throwError = false;
});

it('should get locations from api', async () => {
  const response = await handler(mockAISALBEvent(path));
  assert(response.body);
  const responseObj = await parseResponse(response.body);
  expect(responseObj.length).toBe(3);
  expect(responseObj).toMatchSnapshot();
});

it('should get internal server error when api call fails and no cached response', async () => {
  throwError = true;
  const response = await handler(mockAISALBEvent(path));
  expect(response.statusCode).toBe(503);
});

it('should return right cache headers', async () => {
  const response = await handler(mockAISALBEvent(path));
  assert(response.body);
  const headers = getCacheControlHeaders('aisvessels')?.['Cache-Control'];
  expect(response?.multiValueHeaders?.['Cache-Control']).toStrictEqual(headers);
});
