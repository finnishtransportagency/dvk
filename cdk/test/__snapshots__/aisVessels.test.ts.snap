// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get vessels from api 1`] = `
[
  {
    "callSign": "MC001",
    "destination": "KOTKA",
    "draught": 0.32,
    "eta": "00-00 00:00",
    "imo": 7734789,
    "mmsi": 11122233001,
    "name": "MOCK1",
    "posType": 1,
    "referencePointA": 8,
    "referencePointB": 20,
    "referencePointC": 4,
    "referencePointD": 4,
    "shipType": 52,
    "timestamp": "2022-04-22T18:30:05.774Z",
  },
  {
    "callSign": "MC002",
    "destination": "TURKU",
    "draught": 0.75,
    "eta": "00-00 00:00",
    "imo": 9505508,
    "mmsi": 99988877001,
    "name": "MOCK2",
    "posType": 1,
    "referencePointA": 35,
    "referencePointB": 63,
    "referencePointC": 10,
    "referencePointD": 10,
    "shipType": 90,
    "timestamp": "2022-02-05T22:59:32.418Z",
  },
  {
    "callSign": "MC003",
    "destination": "HELSINKI",
    "draught": 0.5599999999999999,
    "eta": "00-00 00:00",
    "imo": 9684823,
    "mmsi": 66655544001,
    "name": "MOCK3",
    "posType": 3,
    "referencePointA": 12,
    "referencePointB": 12,
    "referencePointC": 6,
    "referencePointD": 5,
    "shipType": 52,
    "timestamp": "2022-08-03T05:11:33.045Z",
  },
]
`;

exports[`should get vessels from api when cache expired 1`] = `
[
  {
    "callSign": "MC001",
    "destination": "KOTKA",
    "draught": 0.32,
    "eta": "00-00 00:00",
    "imo": 7734789,
    "mmsi": 11122233001,
    "name": "MOCK1",
    "posType": 1,
    "referencePointA": 8,
    "referencePointB": 20,
    "referencePointC": 4,
    "referencePointD": 4,
    "shipType": 52,
    "timestamp": "2022-04-22T18:30:05.774Z",
  },
  {
    "callSign": "MC002",
    "destination": "TURKU",
    "draught": 0.75,
    "eta": "00-00 00:00",
    "imo": 9505508,
    "mmsi": 99988877001,
    "name": "MOCK2",
    "posType": 1,
    "referencePointA": 35,
    "referencePointB": 63,
    "referencePointC": 10,
    "referencePointD": 10,
    "shipType": 90,
    "timestamp": "2022-02-05T22:59:32.418Z",
  },
  {
    "callSign": "MC003",
    "destination": "HELSINKI",
    "draught": 0.5599999999999999,
    "eta": "00-00 00:00",
    "imo": 9684823,
    "mmsi": 66655544001,
    "name": "MOCK3",
    "posType": 3,
    "referencePointA": 12,
    "referencePointB": 12,
    "referencePointC": 6,
    "referencePointD": 5,
    "shipType": 52,
    "timestamp": "2022-08-03T05:11:33.045Z",
  },
]
`;

exports[`should get vessels from cache 1`] = `
[
  {
    "callSign": "TS0001",
    "destination": "HANKO",
    "draught": 3.1,
    "eta": "10-24 13:00",
    "imo": 1234567,
    "mmsi": 11122233000,
    "name": "TEST1",
    "posType": 1,
    "referencePointA": 19,
    "referencePointB": 12,
    "referencePointC": 2,
    "referencePointD": 5,
    "shipType": 90,
    "timestamp": "2023-1009T14:06:20.027Z",
  },
  {
    "callSign": "TS0002",
    "destination": "HAMINA",
    "draught": 10.9,
    "eta": "00-00 24:60",
    "imo": 9876543,
    "mmsi": 99988877000,
    "name": "TEST2",
    "posType": 1,
    "referencePointA": 148,
    "referencePointB": 36,
    "referencePointC": 17,
    "referencePointD": 10,
    "shipType": 80,
    "timestamp": "2023-03-03T22:32:22.428Z",
  },
]
`;

exports[`should get vessels from cache when api call fails 1`] = `
[
  {
    "callSign": "TS0001",
    "destination": "HANKO",
    "draught": 3.1,
    "eta": "10-24 13:00",
    "imo": 1234567,
    "mmsi": 11122233000,
    "name": "TEST1",
    "posType": 1,
    "referencePointA": 19,
    "referencePointB": 12,
    "referencePointC": 2,
    "referencePointD": 5,
    "shipType": 90,
    "timestamp": "2023-1009T14:06:20.027Z",
  },
  {
    "callSign": "TS0002",
    "destination": "HAMINA",
    "draught": 10.9,
    "eta": "00-00 24:60",
    "imo": 9876543,
    "mmsi": 99988877000,
    "name": "TEST2",
    "posType": 1,
    "referencePointA": 148,
    "referencePointB": 36,
    "referencePointC": 17,
    "referencePointD": 10,
    "shipType": 80,
    "timestamp": "2023-03-03T22:32:22.428Z",
  },
]
`;
